### PR TITLE
SG-30793 Prioritize User-Restricted Pipeline Configuration

### PR DESCRIPTION
**Pull request by minna-feng**
If there are two site-wide pipeline configurations and a user with restrictions on one pipeline logs in, the pipeline with that user's restrictions will be prioritized.

### CODE REVIEW

**Review by dsauve-adsk — APPROVED** [review]

File: python/shotgun_desktop/paths.py [hunk 1]
```diff
@@ -98,11 +99,19 @@ def get_pipeline_configuration_info(connection):
...
+        # Pick the pipeline configuration with restricted to user, if any.
+        logged_user_id = sgtk.get_authenticated_user().resolve_entity().get("id")
+        for p in pcs:
+            for restricted_user in p.get("users"):
+                if logged_user_id == restricted_user.get("id"):
+                    pc = p
+
+        # If not, pick the last (lowest id).
+        if not pc:
+            pc = pcs[-1]
```
Comment (julien-lang) [inline python/shotgun_desktop/paths.py:113]: I see a few problems here especially when considering the code below (warning if several items are selected).

First, if no configurations (`pcs` items) match our current user, we don't want to select the **last** config of the whole `pcs` list but a subset of `pcs` that only contains configuration with no user restriction.

If we do have a subset of `pcs` with a user restriction that contains `logged_user_id`, we will pick the last one indeed but we need to override `pcs` with this subset to make the code below works.

In other words, you first need to build a subset of `pcs` with the `pc` that contains `logged_user_id`. If not empty, you override `pcs` with it.
Overwise, you need to override `pcs` with a subset of `pc` having no user restrictions.
Then, you pick the last pcs (common to both cases so I believe this part of the code should be unchanged.


Please validate these issues with unit tests first before pushing the fix.

File: python/shotgun_desktop/paths.py [hunk 2]
```diff
@@ -98,11 +99,19 @@ def get_pipeline_configuration_info(connection):
     )
     logger.debug(pprint.pformat(pcs))
 
-    if len(pcs) == 0:
-        pc = None
-    else:
-        # Pick the last result. See the big comment before the Shotgun query to understand.
-        pc = pcs[-1]
+    pc = None
+    if len(pcs) != 0:
```
Comment (julien-lang) [inline python/shotgun_desktop/paths.py:103]: Or > 0. I think it's best to simply use `if len(pcs):`

But I like that you initialize `pc` instead of the if/else.

**Review by julien-lang — CHANGES_REQUESTED** [review]

File: python/shotgun_desktop/paths.py [hunk 3]
```diff
@@ -98,11 +99,28 @@ def get_pipeline_configuration_info(connection):
...
+        logged_user_id = sgtk.get_authenticated_user().resolve_entity().get("id")
+        restricted_pcs = []
+        unrestricted_pcs = []
+        for p in pcs:
+            if not p.get("users"):
+                unrestricted_pcs.append(p)
+                continue
+            for restricted_user in p.get("users"):
+                if logged_user_id == restricted_user.get("id"):
+                    restricted_pcs.append(p)
```
Comment (julien-lang) [inline python/shotgun_desktop/paths.py:114]: You could use a generator to avoid this loop here that make the algorithm look complex:

```python
if not p.get("users"):
    unrestricted_pcs.append(p)
elif logged_user_id in [ u.get("id") for u in p.get("users") ]:
    restricted_pcs.append(p)

```

File: python/shotgun_desktop/paths.py [hunk 4]
```diff
@@ -98,11 +99,29 @@ def get_pipeline_configuration_info(connection):
...
+                unrestricted_pcs.append(p)
+            elif logged_user_id in [u.get("id") for u in p.get("users")]:
+                restricted_pcs.append(p)
+
+        if restricted_pcs:
+            pcs = restricted_pcs
+        elif unrestricted_pcs:
+            pcs = unrestricted_pcs
+        else:
+            pcs = []
```
Comment (julien-lang) [inline python/shotgun_desktop/paths.py:119]: Minor point:
```python
        elif unrestricted_pcs:
            pcs = unrestricted_pcs
        else:
            pcs = []
```

is equivalent and can be simplified to:
```python
        else:
            pcs = unrestricted_pcs
```

**Review by julien-lang — APPROVED** [review]
LGTM!
